### PR TITLE
Derived Loader Fixes

### PR DESCRIPTION
**Pull request by incrypto32**
This PR completes https://github.com/graphprotocol/graph-tooling/pull/1190
- Fixes some issues with Immutable JS while rebasing 
- Fixes wrong type names being generated for derived loaders

### CODE REVIEW

File: packages/ts/index.ts [hunk 1]
```diff
@@ -26,6 +26,7 @@ export declare namespace store {
   /** If the entity was not created in the block, this function will return null. */
   // Matches the host function https://github.com/graphprotocol/graph-node/blob/9f4a1821146b18f6f49165305e9a8c0795120fad/runtime/wasm/src/module/mod.rs#L1091-L1099
   function get_in_block(entity: string, id: string): Entity | null;
+  function loadRelated(entity: string, id: string, field: string): Array<Entity>;
```
Comment (saihaj) [inline packages/ts/index.ts:29]: can you add a changeset for explaining the upgrade to `graph-ts`

File: packages/cli/src/codegen/schema.ts [hunk 2]
```diff
@@ -104,13 +105,31 @@ export default class SchemaCodeGenerator {
       .filter(Boolean) as Array<tsCodegen.Class>;
   }
 
+  generateDerivedLoaders() {
+    const fields = this.schema.ast.definitions
+      .filter(def => this._isEntityTypeDefinition(def))
+      .flatMap((def: any) => def.fields)
+      .filter((def: any) => this._isDerivedField(def))
+      .map((def: FieldDefinitionNode) => this._getTypeNameForField(def.type));
```
Comment (saihaj) [inline packages/cli/src/codegen/schema.ts:113]: let's not use `any` type here. When you filter ` .filter(def => this._isEntityTypeDefinition(def))` then it returns a `ObjectTypeDefinitionNode`  which is more narrowed down type

File: packages/cli/src/codegen/schema.ts [hunk 3]
```diff
@@ -104,13 +105,31 @@ export default class SchemaCodeGenerator {
...
+
+    return [...new Set(fields)].map((typeName: any) => {
+      return this._generateDerivedLoader(typeName);
+    });
```
Comment (saihaj) [inline packages/cli/src/codegen/schema.ts:117]: same here let's avoid `any`'s

File: packages/cli/src/type-generator.ts [hunk 4]
```diff
@@ -161,6 +161,7 @@ export default class TypeGenerator {
             GENERATED_FILE_NOTE,
             ...codeGenerator.generateModuleImports(),
             ...codeGenerator.generateTypes(),
+            ...codeGenerator.generateDerivedLoaders(),
```
Comment (saihaj) [inline packages/cli/src/type-generator.ts:164]: we should add changeset explaining the changes what this feature brings in, so we can have it documented in changelog.

**Review by saihaj — CHANGES_REQUESTED** [review]
looks great in terms of feature. Few nits on code which we should fix and then good to merge

File: packages/cli/src/codegen/schema.ts (hunk 2 above)
Comment (incrypto32) [reply packages/cli/src/codegen/schema.ts]: understood, but removing that any resulted in some weird errors, i have now manually casted these arrays to `ObjectTypeDefinitionNode[]` that seemed to have solved the problem

File: packages/cli/src/type-generator.ts (hunk 4 above)
Comment (incrypto32) [reply packages/cli/src/type-generator.ts:164]: Added changesets

File: packages/cli/src/codegen/schema.ts [hunk 5]
```diff
@@ -104,13 +105,36 @@ export default class SchemaCodeGenerator {
...
+  }
+
   _isEntityTypeDefinition(def: DefinitionNode): def is ObjectTypeDefinitionNode {
     return (
       def.kind === 'ObjectTypeDefinition' &&
       def.directives?.find(directive => directive.name.value === 'entity') !== undefined
     );
   }
 
+  _isDerivedField(field: any): boolean {
```
Comment (saihaj) [inline packages/cli/src/codegen/schema.ts:132]:
```suggestion
  _isDerivedField(field: DefinitionNode): boolean {
```

File: packages/cli/src/codegen/schema.ts [hunk 6]
```diff
@@ -240,7 +326,56 @@ export default class SchemaCodeGenerator {
...
+      return this._returnTypeForDervied(gqlType.type);
+    }
+    if (gqlType.kind === 'ListType') {
+      return this._returnTypeForDervied(gqlType.type);
+    }
+    const type = tsCodegen.namedType(gqlType.name.value + 'Loader');
+    return type;
+  }
+
+  _generatedEntityDerivedFieldGetter(_entityDef: any, fieldDef: FieldDefinitionNode) {
```
Comment (saihaj) [inline packages/cli/src/codegen/schema.ts:355]: we can probably do `ObjectTypeDefinitionNode` here too 🤔

File: packages/cli/src/codegen/schema.ts [hunk 7]
```diff
@@ -240,7 +326,56 @@ export default class SchemaCodeGenerator {
...
+      `get ${name}`,
+      [],
+      returnType,
+      `
+        return new ${returnType}('${entityName}', this.get('id')!.toString(), '${name}')
+      `,
+    );
+  }
+
+  _returnTypeForDervied(gqlType: any): any {
```
Comment (saihaj) [inline packages/cli/src/codegen/schema.ts:344]: need to define the return type

**Review by saihaj — APPROVED** [review]

**Pull request merged**